ord_schema/message_helpers.py

```diff
@@ -338,6 +342,72 @@ def check_compound_identifiers(compound):
         raise ValueError(f'structural identifiers are inconsistent: {smiles}')
 
 
+def get_reaction_smiles(message, validate=True):
+    """Fetches or generates a reaction SMILES.
+
+    Args:
+        message: reaction_pb2.Reaction message.
+        validate: Boolean whether to validate the reaction SMILES with rdkit.
+
+    Returns:
+        Text reaction SMILES.
+
+    Raises:
+        ValueError: If the reaction contains errors.
+    """
+    for identifier in message.identifiers:
+        if identifier.type == reaction_pb2.ReactionIdentifier.REACTION_SMILES:
+            return identifier.value
+    reactants, agents, products = set(), set(), set()
+    roles = reaction_pb2.Compound.ReactionRole()
+    for key in sorted(message.inputs):
+        for compound in message.inputs[key].components:
+            smiles = smiles_from_compound(compound)
```

Comment:
I'm interpreting this as follows: a ValueError will be raised if _any_ compound cannot be resolved to a SMILES string. Suppose a reaction takes place in "hexanes" (mixture of isomers) as a solvent. The reactants, reagents, etc. can be well-defined but the solvent might be defined by name only. Do we want to still raise an error? What if a catalyst is "Pd/C" (super common) or a biological enzyme so there is no SMILES string and it is only defined by name?

Reply:
If I'm correct in that `smiles_from_compound` will raise an error, perhaps we should have some kwarg that is akin to `reactants_only` or `include_agents`? Alternatively, we could have some wrapper that can issue warnings instead of errors

Reply:
My plan was to catch the error as part of build_database.py and just not generate a reaction SMILES in those cases. I think this is ok since it's just for postgres---users could still search by the inputs/outputs of a reaction without searching based on the entire thing. I'm not sure what to do about `NAME`d components yet other than adding more columns for these that people can search against.

That being said, if you'd prefer to generate "incomplete" reaction SMILES that would be fine with me...what do you think? (This affects several other comments as well.)

Reply:
I do think it would be worth allowing incomplete reaction SMILES in the database

Reply:
Done. PTAL.